### ISSUE TITLE
from_str should reject extra characters before or after duration

### ISSUE THREAD

**Issue by benhoyt** [icholy/durationpy#11]
Currently `from_str` accepts inputs with extra characters before or after the duration string, like this:

```
>>> from_str('X72h3m0.5msY')
datetime.timedelta(days=3, seconds=180, microseconds=500)
```

It seems to me this should be a DurationError, as it's not a valid duration. (For comparison, Go's `time.ParseDuration` rejects this: https://go.dev/play/p/2mN7T4Vdrxx)

One simple way to do this is to use `pattern.finditer` instead of `pattern.findall`. In my own variant of this function I do that like this:

```python
    matches = list(_DURATION_RE.finditer(s))
    if not matches:
        raise ValueError('invalid duration: no number-unit groups')
    if matches[0].start() != 0 or matches[-1].end() != len(s):
        raise ValueError('invalid duration: extra input at start or end')
    ...
    for match in matches:
        value, unit = match.groups()
```

**Comment by icholy** [comment]
@benhoyt thanks for the report, the fix looks good. Let me know if you want to open a PR. I can do it otherwise.

**Comment by benhoyt** [comment]
I've submitted a PR to fix this, #12. I've addressed the sign issue by moving the sign handling to before the regex matching.
